k9mail/src/main/java/com/fsck/k9/activity/setup/AccountSetupPresenter.java

```diff
@@ -243,14 +280,89 @@ protected ProviderInfo doInBackground(Void... params) {
                 if (providerInfo != null) return providerInfo;
 
                 providerInfo = autodiscover.findProviderInfo(email);
+                if (providerInfo != null) return providerInfo;
 
-                return providerInfo;
+                testDomain(domain);
+
+                return null;
+            }
+
+            private void testDomain(String domain) {
```

Comment:
I feel like there is a lot of logic here that warrants its own class. There are already `Autoconfigure*` classes, why not have an `AutoconfigureHeuristic` one?